docs/docs/writing-plugins/the-rules-api/concepts.mdx

```diff
@@ -23,176 +23,251 @@ async def int_to_str(i: int) -> str:
     return str(i)
 ```
 
-Although any Python type, including builtin types like `int`, can be a parameter or return type of a rule, in almost all cases rules will deal with values of custom Python classes.
+Rules are typically module-level functions. In some cases you can define rules in nested scopes,  such as inside a class or function body. But this is useful only in specific, special cases in the Pants codebase, and you are unlikely to need to use this in practice.
 
-Generally, rules correspond to a step in your build process. For example, when adding a new linter, you may have a rule that maps `(Target, Shellcheck) -> LintResult`:
+Although any immutable Python type, including builtin types like `int`, can be a parameter or a return type of a rule, in almost all cases rules will deal with values of custom Python classes. These are are typically implemented as [frozen dataclasses](https://docs.python.org/3/library/dataclasses.html).
```

Comment:
> These are are typically implemented as [frozen dataclasses]

Any value in adding why they're this way? Why frozen, why a dataclass, etc? I've always just cargo-culted this, and even in some of the base classes - we say "Subclasses must set `@dataclass(frozen=True)` for their declared fields to be recognized.". Never really gave it much thought, but I miss adding them quite often.

Reply:
This might also be addressed below tangentially? 
> The engine invokes rules concurrently where possible, to make use of all available local and remote cores. This is why rule params and return values must be immutable.

Reply:
Also, mentioned before in the Dataclasses section. Another case of maybe foreshadow the why - as you have to read pretty far down to see (or even just a link anchor?)